magento/loaders/productDetailsPage.ts

```diff
@@ -0,0 +1,168 @@
+import type { ListItem, ProductDetailsPage } from "../../commerce/types.ts";
+import type { RequestURLParam } from "../../website/functions/requestToParam.ts";
+import { AppContext } from "../mod.ts";
+import { KEY_FIELD, KEY_VALUE, URL_KEY } from "../utils/constants.ts";
+import { toProduct } from "../utils/transform.ts";
+
+export interface Props {
+  slug: RequestURLParam;
+  currencyCode?: string;
+}
+
+/**
+ * @title VNDA Integration
+ * @description Product Details Page loader
+ */
+async function loader(
+  props: Props,
+  req: Request,
+  ctx: AppContext,
+): Promise<ProductDetailsPage | null> {
+  console.log("loader start");
+  const url = new URL(req.url);
+  const { slug, currencyCode = "" } = props;
+  const { clientGuest, clientAdmin, site, storeId } = ctx;
+
+  if (!slug) {
+    return null;
+  }
+
+  const getMaybeProduct = async (slug: string) => {
+    try {
+      const response = await clientAdmin["GET /rest/:site/V1/products"]({
+        site,
+        currencyCode: currencyCode,
+        [KEY_FIELD]: URL_KEY,
+        [KEY_VALUE]: slug,
+      }).then((res) => res.json());
+      return response.items[0];
+    } catch (_error) {
+      return null;
+    }
+  };
+  const getProductPriceAndImages = async (sku: string | null) => {
+    if (!sku) {
+      return null;
+    }
+    try {
+      const response = await clientGuest
+        ["GET /rest/:site/V1/products-render-info"]({
+          site,
+          [KEY_FIELD]: "sku",
+          [KEY_VALUE]: sku,
+          storeId: storeId,
+          currencyCode: currencyCode,
+          fields: "items[price_info,image,currency_code,url]",
+        }).then((res) => res.json());
+
+      return {
+        priceInfo: response.items[0].price_info,
+        currencyCode: response.items[0].currency_code,
+        images: response.items[0].images,
+        url: response.items[0].url,
+      };
+    } catch (_error) {
+      return null;
+    }
+  };
```

Comment:
Essa estrutura não esta legal, esta limitando muito o que voce pode ou não usar das APIs. O melhor cenario, permitindo que isso seja uma estrutura mais flexivel e que permita que o loader possa crescer com maior facilidade é:
- Chamar as APIs de uma vez com Promise.all
- Transformar os dados de uma vez, sem separar "o que é importante": 

```js
toProduct(api1, api2, api3)
```

Voce pode tratar os dados se necessário como o `pickSku` no loader da VTEX, mas não limitar o que vai ser enviado no toProduct

Reply:
@carcara-gustavo e que nessa função só retorna isso mesmo, eu só disse oque retorna para facilitar na tipagem se eu tirar ele fala que tem tudo que tem em um produto, posso criar uma tipagem diferente mas cai no mesmo problema talvez 

- primeiro ponto: tudo que dar eu to chamando no promisse.all mesmo, do 1 chamada que não da todas as outras depende dela
- segundo ponto: realmente da a impressão que to transformando o dado porem to simplificando o retorno da função vou pensar em como melhorar isso